### DISCUSSION TITLE
SNN model for load forecasting using Lava framework implementation on Loihi 2

### DISCUSSION THREAD

**Discussion by megalover** [lava-nc/lava-dl#84] [General]
@mgkwill  
I am working on a project about creating an SNN model for load forecasting (Time-series data) on the Lava framework. However, I am struggling to implement SNN with time-series data. 
Could you give a simple example to use the SNN function of time series? I apologize for this may disturb you, I am a green hand.

In any case, thanks for your help.

**Comment by bamsumit** [comment]
Of the available tutorials, [PilotNet steering angle prediction](https://github.com/lava-nc/lava-dl/blob/main/tutorials/lava/lib/dl/slayer/pilotnet/train.ipynb) example is the closest. Check it out. If that does not help, describe your use case and network architecture and I am happy to guide.

**Reply by megalover** [reply]
@bamsumit Thank you so much for your response. Actually, I have already tried to compile this Pilotnet example on my load forecasting time-series dataset, but my data is just a vale of Megawatt at each time point and I can create it in 2D (number of steps,1) for my input and the target is the value of next time point, so I cannot figure out how to apply it in this Pilonet dataset that has like 5 dimensions (5D) of data. I have already tried but my data dimensions are still mismatched with the Convolution part of this Network in the example. Do you have any guide for me, please?
[Energy_Time_Series_Prediction_SDNN.zip](https://github.com/lava-nc/lava-dl/files/9123335/Energy_Time_Series_Prediction_SDNN.zip)

**Reply by bamsumit** [reply]
It sounds like your data is of dimension `(1, num_steps)` (In lava-dl, we always assume time is the last dimension). Typically when people do time series prediction, they take the last `n` timesteps and predict the next `n+1`th time step. So you could arrange your data in the following form
```
[[x[n  ], x[n+1], ..., x[T-1  ]],
 [x[n-1], x[n  ], ..., x[T-2  ]],
 ...
 [x[0  ], x[1  ], ..., x[T-n-1]]
```
and use either __dense__ or __1d conv__ (you will need to reshape your `(n, T-n)` data to `(1, 1, n, T-n)` to use 1d conv) to process the data.

**Reply by megalover** [reply]
@bamsumit Thank you so much I really appreciate your help, I can now implement my SNN model for time-series data. Then, I would like to run the model created on Loihi 2 via cloud-hosted Loihi systems from the Intel Neuromorphic Research Community (INRC). Could you please let me know how can I access to that systems or hardware? Thanks

**Comment by bamsumit** [comment]
https://intel-ncl.atlassian.net/wiki/spaces/INRC/pages/1784807425/Join+the+INRC

**Comment by megalover** [comment]
Hello @bamsumit , I have a question about the slayer block in lava-dl. The sigma-delta blocks for conv, pool, etc. are the **direct training SNN** or **CNN to SNN conversion** method? I'm not sure about block class in lava-dl. Could you please clarify this? Thank you.

**Reply by bamsumit** [reply]
@megalover everything under `lava.lib.dl.slayer` are for **direct training** and everything under `lava.lib.dl.bootstrap` are for **ANN to SNN conversion**. For sigma-delta, we only support direct training. It's only available in `lava.lib.dl.slayer.blocks.sigma_delta`.

**Reply by megalover** [reply]
@bamsumit Thank you. Then, the learning rule on this Pilotnet tutorial is the Spatio-Temporal Back-Propagation (STBP), or Spikeprop, or just common error backpropagation or something? And do you know why this time series data doesn't work for the other neuron, like CUBA and ALIF? When I try, the other neurons seem not to fire, and the model doesn't train, but I don't know why. Thank you so much for your help.
